File: cpp/triangle/Triangle.cpp
#include "Triangle.h"

bool isEqualDbl(double a, double b) {
        return fabs(a - b) < DBL_EPSILON;
}

bool isValidTriangle(double n1, double n2, double n3) {
        if (n1 <= 0 || n2 <= 0 || n3 <= 0) {
                return false; 
        }

        if (n1 + n2 < n3 || n1 + n3 < n2 || n2 + n3 < n1) {
                return false; 
        }

        return true; 
}

int triangle::kind(double n1, double n2, double n3) 
{
        if (!isValidTriangle(n1, n2, n3)) {
                throw std::domain_error("Invalid Triangle!");
        }

        if (isEqualDbl(n1, n2) && isEqualDbl(n2, n3)) {
                return equilateral;
        }
        else if (isEqualDbl(n1, n2) || isEqualDbl(n1, n3) || isEqualDbl(n2, n3)) {
                return isosceles; 
        }

        return scalene; 
}

File: php/triangle/triangle.php
<?php

class Triangle 
{
    private $a;
    private $b; 
    private $c; 

    function __construct($a, $b, $c) {
        $this->a = $a; 
        $this->b = $b;
        $this->c = $c; 
    }

    function kind() {
        if ($this->a <= 0 || $this->b <= 0 || $this->c <= 0) {
            throw new Exception("Sides of negative length or length 0 are not allowed.");
        }

        $sides = array($this->a, $this->b, $this->c);
        sort($sides); 
        if ($sides[2] >= $sides[0] + $sides[1]) {
            throw new Exception("This is not a valid triangle.");
        }

        if ($sides[0] === $sides[2]) {
            return "equilateral";
        }

        if ($sides[0] === $sides[1] || $sides[1] === $sides[2]) {
            return "isosceles";
        }

        return "scalene";
    }
}